### ISSUE TITLE
Allocator very slow to reclaim memory after running for sufficiently long

### ISSUE THREAD

**Issue by aterenin** [JuliaGPU/CUDA.jl#137]
This is related to the Flux issue https://github.com/FluxML/Flux.jl/issues/736 which after playing around with in an interactive session, I am becoming increasingly convinced is related to strange behavior in the CuArrays memory allocator. Self-contained MWE can be found in that issue. 

The problem is: *after running sufficiently long*, buffers do not get reclaimed by CuArrays quickly enough, which causes the GPU to run out of memory, and performance to slow to a crawl.

This is not a `Tracker` issue - I've replicated it with no `TrackedArray` types. It occurs when training MNIST with a completely standard static graph deep convolutional network with no fancy components. 

Example output *after just starting Julia*.

```
finished minibatch 5
Total GPU memory usage: 30.0% (2.363 GiB/7.766 GiB)
CuArrays.jl pool usage: 48.0% (357.994 MiB in use by 4302 buffer(s), 810.212 MiB idle)
finished minibatch 10
Total GPU memory usage: 31.0% (2.394 GiB/7.766 GiB)
CuArrays.jl pool usage: 49.0% (537.099 MiB in use by 5007 buffer(s), 662.265 MiB idle)
finished minibatch 15
Total GPU memory usage: 35.0% (2.711 GiB/7.766 GiB)
CuArrays.jl pool usage: 55.00000000000001% (1.489 GiB in use by 8782 buffer(s), 0 bytes idle)
finished minibatch 20
Total GPU memory usage: 41.0% (3.146 GiB/7.766 GiB)
CuArrays.jl pool usage: 61.0% (932.075 MiB in use by 6517 buffer(s), 1.014 GiB idle)
finished minibatch 25
Total GPU memory usage: 41.0% (3.146 GiB/7.766 GiB)
CuArrays.jl pool usage: 61.0% (1.875 GiB in use by 10292 buffer(s), 51.035 MiB idle)
 ────────────────────────────────────────────────
                                   Time          
                           ──────────────────────
     Tot / % measured:          63.5s / 0.89%    

 Section           ncalls     time   %tot     avg
 ────────────────────────────────────────────────
 pooled alloc       80.2k    480ms  85.3%  5.99μs
   1 try alloc      10.9k    469ms  83.3%  42.9μs
 background task       11   92.8ms  16.5%  8.43ms
   reclaim             11   17.1μs  0.00%  1.55μs
   scan                11   16.6μs  0.00%  1.51μs
 ────────────────────────────────────────────────
```

Output *after running many epochs of Flux training* - notice how memory usage continues to grow. Timings are after 100 minibatches.
```
finished minibatch 5
Total GPU memory usage: 100.0% (7.763 GiB/7.766 GiB)
CuArrays.jl pool usage: 83.0% (740.112 MiB in use by 5811 buffer(s), 5.730 GiB idle)
finished minibatch 10
Total GPU memory usage: 100.0% (7.763 GiB/7.766 GiB)
CuArrays.jl pool usage: 83.0% (1.687 GiB in use by 9586 buffer(s), 4.765 GiB idle)
finished minibatch 15
Total GPU memory usage: 100.0% (7.763 GiB/7.766 GiB)
CuArrays.jl pool usage: 83.0% (2.651 GiB in use by 13361 buffer(s), 3.801 GiB idle)
finished minibatch 20
Total GPU memory usage: 100.0% (7.763 GiB/7.766 GiB)
CuArrays.jl pool usage: 83.0% (3.616 GiB in use by 17136 buffer(s), 2.837 GiB idle)
finished minibatch 25
Total GPU memory usage: 100.0% (7.763 GiB/7.766 GiB)
CuArrays.jl pool usage: 83.0% (4.580 GiB in use by 20911 buffer(s), 1.872 GiB idle)
 ────────────────────────────────────────────────
                                   Time          
                           ──────────────────────
     Tot / % measured:          1385s / 42.4%    

 Section           ncalls     time   %tot     avg
 ────────────────────────────────────────────────
 pooled alloc       28.4M     295s  50.1%  10.4μs
   2 gc(false)        331     289s  49.2%   874ms
   1 try alloc      33.4k    1.70s  0.29%  50.8μs
 background task      885     294s  50.0%   332ms
   reclaim            885   49.7ms  0.01%  56.2μs
   scan               885    731μs  0.00%   826ns
 ────────────────────────────────────────────────
```

Output *if `GC.gc()` is added after every 5 minibatches*.
```
finished minibatch 5
Total GPU memory usage: 27.0% (2.086 GiB/7.766 GiB)
CuArrays.jl pool usage: 38.0% (255.660 MiB in use by 3967 buffer(s), 563.883 MiB idle)
finished minibatch 10
Total GPU memory usage: 32.0% (2.498 GiB/7.766 GiB)
CuArrays.jl pool usage: 49.0% (255.660 MiB in use by 3967 buffer(s), 987.439 MiB idle)
finished minibatch 15
Total GPU memory usage: 32.0% (2.498 GiB/7.766 GiB)
CuArrays.jl pool usage: 49.0% (255.660 MiB in use by 3967 buffer(s), 987.439 MiB idle)
finished minibatch 20
Total GPU memory usage: 32.0% (2.498 GiB/7.766 GiB)
CuArrays.jl pool usage: 49.0% (255.660 MiB in use by 3967 buffer(s), 987.439 MiB idle)
finished minibatch 25
Total GPU memory usage: 32.0% (2.498 GiB/7.766 GiB)
CuArrays.jl pool usage: 49.0% (255.660 MiB in use by 3967 buffer(s), 987.439 MiB idle)
```

I'd like to ask for some help in debugging this further. What can I do to further try to find out what is the issue?

**Comment by maleadt** [comment]
I couldn't reproduce the slowdown with the following example from somewhere in this issue:

```julia
using CUDA, Flux, IterTools

function make_model1()
    first_layer = Conv((1, 1), 1=>16)

    # performance becomes very bad during the 8th epoch
    layers = [Conv((3, 3), 16=>16, pad=(1, 1)) for _ in 1:50]
    last_layer = Conv((1, 1), 16=>1)

    Chain(first_layer, layers..., last_layer) |> gpu
end

function make_model2()
    # performance slowly becomes worse, but not suddenly so
    layers = [Conv((3, 3), 1=>1, pad=(1, 1)) for _ in 1:50]

    Chain(layers...) |> gpu
end

function test_training()
    model = make_model2()

    data = collect(repeatedly(() -> (CUDA.randn(200, 200, 1, 1), CUDA.randn(200, 200, 1, 1)), 2500))

    loss(x, y) = sum(model(y) - x)
    opt = ADAM()
    ps = params(model)

    epoch = 1
    while true
        @info "Epoch $epoch"
        @time Flux.train!(loss, ps, data, opt)
        epoch += 1
    end
end

test_training()
```

Both using the binned allocator and the new CUDA-based allocator, epoch timings remained identical, so it seems like some improvements to the allocator or to Flux.jl may have fixed this case? I did run on a GPU with a fair amount of memory though. Regardless, the new allocator used on CUDA 11.2 (only available on CUDA#master for now) should also improve performance when there's a lot of memory pressure. In the case that there isn't a lot of memory pressure, the new pool slightly degrades performance, but (1) that seems worth it, and (2) the memory pool is expected to improve still. Some detailed timings:

Binned pool (CUDA <11.2)

```
 ────────────────────────────────────────────────
                                   Time          
                           ──────────────────────
     Tot / % measured:          1018s / 1.87%    

 Section           ncalls     time   %tot     avg
 ────────────────────────────────────────────────
 pooled alloc       15.7M    16.2s  85.2%  1.03μs
   0. repopulate    2.23M    4.50s  23.6%  2.02μs
     pooled free    46.8k   7.65ms  0.04%   164ns
   1. try alloc     12.7k    230ms  1.20%  18.1μs
   pooled free      9.40k   1.54ms  0.01%   164ns
 pooled free        15.7M    2.83s  14.8%   181ns
 ────────────────────────────────────────────────
Allocator timings:
 ────────────────────────────────────────
                           Time          
                   ──────────────────────
 Tot / % measured:      1018s / 0.02%    

 Section   ncalls     time   %tot     avg
 ────────────────────────────────────────
 alloc      12.7k    195ms   100%  15.3μs
 ────────────────────────────────────────
```

CUDA pool (CUDA 11.2+)

```
Memory pool timings:
 ────────────────────────────────────────────────
                                   Time          
                           ──────────────────────
     Tot / % measured:          1353s / 8.48%    

 Section           ncalls     time   %tot     avg
 ────────────────────────────────────────────────
 pooled alloc       21.0M    85.5s  74.6%  4.07μs
   1.1 alloc        21.0M    66.7s  58.1%  3.17μs
     pooled free     259k    330ms  0.29%  1.28μs
   pooled free       317k    403ms  0.35%  1.27μs
 pooled free        20.4M    29.1s  25.4%  1.43μs
 ────────────────────────────────────────────────
Allocator timings:
 ────────────────────────────────────────
                           Time          
                   ──────────────────────
 Tot / % measured:      1353s / 5.45%    

 Section   ncalls     time   %tot     avg
 ────────────────────────────────────────
 alloc      21.0M    52.5s  71.2%  2.50μs
   free      105k   98.1ms  0.13%   931ns
 free       20.9M    21.2s  28.8%  1.02μs
 ────────────────────────────────────────
```

With improvements from https://github.com/JuliaGPU/CUDA.jl/pull/745:

```
Memory pool timings:
 ────────────────────────────────────────────────
                                   Time          
                           ──────────────────────
     Tot / % measured:           728s / 5.12%    

 Section           ncalls     time   %tot     avg
 ────────────────────────────────────────────────
 pooled alloc       7.61M    25.5s  68.5%  3.35μs
   1.1 alloc        7.61M    17.9s  48.1%  2.35μs
     pooled free    45.9k   67.0ms  0.18%  1.46μs
   pooled free       114k    166ms  0.45%  1.45μs
 pooled free        7.45M    11.7s  31.5%  1.57μs
 ────────────────────────────────────────────────
Allocator timings:
 ────────────────────────────────────────
                           Time          
                   ──────────────────────
 Tot / % measured:       728s / 2.86%    

 Section   ncalls     time   %tot     avg
 ────────────────────────────────────────
 alloc      7.61M    12.1s  58.1%  1.59μs
 free       7.61M    8.71s  41.9%  1.15μs
 ────────────────────────────────────────
```

Bottom line, I think we can close this for now. There's been a lot of improvements, so please test your workloads and let me know how the new allocator performs. Flux still allocates _a lot_ though, 21M GPU allocations in 1500s is always going to take some time.
